### PR TITLE
task/GP-111-Add_Makefile

### PR DESCRIPTION
**Pull request by adrienbesnard-ultra**
Add a Makefile which may used to hide the underlying project management system.

### CODE REVIEW

File: Makefile [hunk 1]
```diff
@@ -0,0 +1,25 @@
...
+SHELL = /usr/bin/env
+.SHELLFLAGS = bash -euo "pipefail" -c
+
+.PHONY: help
+help:
+	@grep -E '^[a-zA-Z_-]+:.*?## .*$$' $(MAKEFILE_LIST) | sort | awk 'BEGIN {FS = ":.*?## "}; {printf "\033[36m%-30s\033[0m %s\n", $$1, $$2}'
+
+.PHONY: test
+test: ## Run tests
+	./gralew test
```
Comment (romain-gilles-ultra) [inline Makefile:12]: `./gralew` does not exits

File: Makefile [hunk 2]
```diff
@@ -0,0 +1,25 @@
...
+
+.PHONY: check
+check: ## Ensure test coverage follow required standards
+	./gralew check
```
Comment (romain-gilles-ultra) [inline Makefile:16]: `./gralew` does not exits

File: Makefile [hunk 3]
```diff
@@ -0,0 +1,25 @@
...
+
+.PHONY: build
+build: ## Build the JAR locally
+	./gralew build
```
Comment (romain-gilles-ultra) [inline Makefile:20]: `./gralew` does not exits

File: Makefile [hunk 4]
```diff
@@ -0,0 +1,25 @@
...
+
+.PHONY: publish
+publish: ## Publish the JAR to GitLab
+	@if [[ -z "$(VERSION)" ]]; then echo -e "\033[0;31mVERSION should be defined\033[0m" && false ; fi
+	./gralew -Pversion="$(VERSION)" publish
```
Comment (romain-gilles-ultra) [inline Makefile:25]: `./gralew` does not exits

File: Makefile (hunk 2 above)
Comment (adrienbesnard-ultra) [reply Makefile]: Fixed

**Comment by romain-gilles-ultra** [comment]
I think we should update the `README` file but you must wait for my PR for that because I updated the `README` a lot: https://github.com/ultraio/gradle-java-template/pull/2

File: Makefile [hunk 5]
```diff
@@ -0,0 +1,27 @@
...
+.PHONY: check
+check: ## Ensure test coverage follow required standards
+	$(GRADLE) check
+
+.PHONY: build
+build: ## Build the JAR locally
+	$(GRADLE) build
+
+.PHONY: publish
+publish: ## Publish the JAR to GitLab
```
Comment (romain-gilles-ultra) [inline Makefile:25]: Rename it to `deploy` as it is the name used in TS projects

Comment (adrienbesnard-ultra) [reply Makefile:25]: Done!

**Comment by damienleger** [comment]
Makefile is lgtm BUT a behavior of `gradlew` is bothering me: the wrapper is supposed to download `gradle` only is there is none on the system (after this https://docs.gradle.org/current/userguide/gradle_wrapper.html#sec:using_wrapper). But it's not the case!

image

**Comment by adrienbesnard-ultra** [comment]
@damienleger I may be wrong, but the `gradlew` wrapper is bind to a specific version: if the wrapper version does not match the system version, then it downloads the binaries. I'm testing it right now just to be sure.

**Comment by adrienbesnard-ultra** [comment]
...And my bad, I was wrong: it downloads it anyway, but once and for all in the `~/.gradle/cache`. I can make a check on the `Makefile` in order to use the system installation, but I'm not sure that it's a good idea as the version may differ and we may have different behavior between devs and CI, for example.

**Comment by romain-gilles-ultra** [comment]
Maybe we can update the version of Gradle in the template for now

**Comment by adrienbesnard-ultra** [comment]
We can, but the wrapper will download the binaries anyway: the system
binaries won't be used. @damienleger: what do you want exactly? That we use
the system binaries?

File: README.md [hunk 6]
```diff
@@ -12,38 +12,38 @@ A description of your project follows. A good description is clear, short, and t
...
+$ .make build
 ```
 
 ###### *The build will fail if coverage criterion are not met*
 
-### Publish a release on gitlab 
+### Publish a release on gitlab
 ```shell script
-$ ./gradlew -Pversion=x.y.z publish
+$ make publish VERSION=x.y.z
```
Comment (romain-gilles-ultra) [inline README.md:46]: I think now `publish` is `deploy`

Comment (adrienbesnard-ultra) [reply README.md:46]: You're right, my bad. Fixed, now.

Comment (adrienbesnard-ultra) [reply README.md]: I think EOL are because of Atom. I'll try to reverse that

**Comment by damienleger** [comment]
> We can, but the wrapper will download the binaries anyway: the system binaries won't be used. @damienleger: what do you want exactly? That we use the system binaries?

The thing is in the CI/CD, build & publish is done is separate job, ie. separate docker image. Download several times Gradle doesn't make more sense. It's better to use a pre-made docker image with Gradle. But in that case I would like gradlew to use the pre-installed Gradle, not download another one. Otherwise the Makefile doesn't make sense to me and then I will invoke Gradle command directly.

**Comment by damienleger** [comment]
I made an update to use gradlew only if gradle not present. Test here:
image

**Comment by adrienbesnard-ultra** [comment]
> I made an update to use gradlew only if gradle not present. Test here:
> image

Haha I stashed something similar for my next PR, but you did it!

**Pull request merged**